[ML] Closing an anomaly detection job now automatically stops its datafeed if necessary

droberts195 opened the pull request:
Previously it was a requirement of the close job API that if the
job had an associated datafeed that that datafeed was stopped
before the job could be closed. Experience has shown that this
is just a pedantic nuisance. If a user closes the job without
first stopping the datafeed then it's just a mistake, and they
then have to make two further calls, to stop the datafeed and
then attempt to close the job again.

This PR changes the behaviour so that if you ask to close a job
whose datafeed is running then the datafeed gets stopped first
as part of the same call. Datafeeds are stopped with the same
level of force as the job close request specified.

droberts195 committed:
[ML] Change close job so that it automatically stops the datafeed

Previously it was a requirement of the close job API that if the
job had an associated datafeed that that datafeed was stopped
before the job could be closed. Experience has shown that this
is just a pedantic nuisance. If a user closes the job without
first stopping the datafeed then it's just a mistake, and they
then have to make two further calls, to stop the datafeed and
then attempt to close the job again.

This PR changes the behaviour so that if you ask to close a job
whose datafeed is running then the datafeed gets stopped first
as part of the same call. Datafeeds are stopped with the same
level of force as the job close request specified.

elasticmachine commented:
Pinging @elastic/ml-core (Team:ML)

benwtrent commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportCloseJobAction.java:
```diff
@@ -204,48 +214,51 @@ protected void doExecute(Task task, CloseJobAction.Request request, ActionListen
+                            ClientHelper.ML_ORIGIN,
+                            StopDatafeedAction.INSTANCE,
+                            request,
+                            ActionListener.wrap(r -> listener.onResponse(r.isStopped()), listener::onFailure));
```
Do we know if the failure here will mention the datafeeds that failed to stop?

It might be useful to indicate "Hey, we didn't close the job becuase its datafeeds were running and we couldn't stop it". This may already be the case, just double checking.

droberts195 replied:
Good point.  I'll improve the error tomorrow.

benwtrent reviewed (approved):
Huge QoL improvement!

Minor comment around error handling on failing to stop datafeeds.

droberts195 commented:
I just realised there's a complication here.  When a datafeed that doesn't have an end time is stopped it auto-closes its job _gracefully_.  This will be problematic in the case where a user _force_ closes a job that has a datafeed running that has an end time.  (I.e., the user force closes a job that's running through a long historical data analysis.) As the changes stand at the moment, the close job action will force stop the datafeed, which will gracefully close the job.  This could take a long time if the model state is large.  Then the force close will operate, find the job is already closed, and do nothing.  So the end result is OK, but the time taken is inconsistent with force close.

droberts195 committed:
Solve the circular close-stop-close problem between job and datafeed

Stop datafeed has always auto-closed the job if the datafeed had
no end time. It _always_ does this gracefully. Now that close job
auto-stops the datafeed, there's a problem if someone force-closes
a job with a started datafeed: it would lead to the job being
_gracefully_ closed.

This change avoids the problem by isolating datafeeds before
force-stopping them in the event of a force-close.

droberts195 committed:
Merge branch 'master' into make_close_job_stop_datafeed

droberts195 commented:
> I just realised there's a complication here.

I've attempted to fix that problem by isolating datafeeds before force-stopping them if they're being stopped as the result of a force-close request.

droberts195 commented:
Jenkins run elasticsearch-ci/docs

benwtrent commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportCloseJobAction.java:
```diff
@@ -249,21 +250,97 @@ void stopDatafeedsIfNecessary(OpenAndClosingIds jobIds, boolean isForce, TimeVal
-                            request,
-                            ActionListener.wrap(r -> listener.onResponse(r.isStopped()), listener::onFailure));
+                        if (isForce) {
+                            // A datafeed with an end time will gracefully close its job when it stops even if it was force stopped.
```
> A datafeed with an end time will gracefully close its job when it stops even if it was force stopped.

This feels like a weird bug to me to begin with. It seems like a weird artifact of having them separate concepts.

Nothing to change, just an observation.

droberts195 replied:
Yes, I think this was unintentional.  The interactions between jobs and datafeeds are so complex that nobody understood the effect of having the datafeed close the job during its stop sequence.  If we'd realised the interactions we'd have probably made it so that a force-stopped datafeed with no end time force-closed its associated job.  However, that would be a breaking change now.  We should consider doing it for 8.0 _if_ the job/datafeed combination project hasn't got as far as combining the persistent tasks by then.

benwtrent commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportCloseJobAction.java:
```diff
@@ -249,21 +250,97 @@ void stopDatafeedsIfNecessary(OpenAndClosingIds jobIds, boolean isForce, TimeVal
+                            // on this.
+                            isolateDatafeeds(jobIds.openJobIds, runningDatafeedIds, ActionListener.wrap(
+                                r -> stopDatafeeds(runningDatafeedIds, true, timeout, listener),
+                                listener::onFailure
```
This is never called, looking at the isolateDatafeeds function, it never calls listener.onFailure. Might be good to make that clear.

benwtrent commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportCloseJobAction.java:
```diff
@@ -249,21 +250,97 @@ void stopDatafeedsIfNecessary(OpenAndClosingIds jobIds, boolean isForce, TimeVal
+            request,
+            ActionListener.wrap(
+                r -> listener.onResponse(r.isStopped()),
+                e -> listener.onFailure(ExceptionsHelper.conflictStatusException(
+                    "failed to close jobs as one or more had started datafeeds that could not be stopped: " +
+                        "started datafeeds [{}], error stopping them [{}]", e, request.getDatafeedId(), e.getMessage())
+                )
```
It would be awesome to one day have the list of datafeeds that were successfully stopped so that those jobs can continue to close.

one day :D

droberts195 replied:
Maybe, but there aren't many reasons why a datafeed would fail to stop to be honest.

What's in this PR is no worse than what we had before, where if you asked to close 100 jobs and one of them had a started datafeed then the whole close request got rejected.  Now we're in the situation where if you ask to close 100 jobs and one of them has a started datafeed and that fails to stop then none of the 100 jobs get closed.

When we eventually reach the stage of the job and datafeed being combined things will naturally get better

benwtrent commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportCloseJobAction.java:
```diff
@@ -249,21 +250,97 @@ void stopDatafeedsIfNecessary(OpenAndClosingIds jobIds, boolean isForce, TimeVal
+        assert numberOfDatafeeds > 0;
+        final AtomicInteger counter = new AtomicInteger();
+
+        for (String runningDatafeedId : runningDatafeedIds) {
```
Seems like a good use of `GroupedActionListener` though it may lose the `info` logging (if that is super important or not).
```
ActionListener<?> groupedActionListener = new GroupedActionListener<?>(ActionListenener.wrap(r -> ...., e->...);
for (String runningDatafeedId : runningDatafeedIds) {
...
ClientHelper.executeAsyncWithOrigin(
                client,
                ClientHelper.ML_ORIGIN,
                IsolateDatafeedAction.INSTANCE,
                request,
                groupedActionListener);
}
```

droberts195 replied:
> though it may lose the info logging (if that is super important or not).

It doesn't completely lose it.  It means there is only one message instead of many in the case where multiple datafeeds cannot be isolated, and the datafeed IDs are only mentioned in the exceptions then, and not in the main message.  Since this will be such a super-rare occurrence and it will be us interpreting the logs I don't think it really matters.

benwtrent commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportIsolateDatafeedAction.java:
```diff
@@ -52,22 +52,26 @@ protected void doExecute(Task task, IsolateDatafeedAction.Request request, Actio
                                  ActionListener<IsolateDatafeedAction.Response> listener) {
         datafeedTask.isolate();
-        listener.onResponse(new IsolateDatafeedAction.Response(false));
+        listener.onResponse(new IsolateDatafeedAction.Response(true));
```
seems weird to make this `true`, but the final response to be `false`. Seems like an arbitrary change.

droberts195 replied:
Previously the response was _always_ `false`, which was a bit crazy.

> but the final response to be `false`

The final response _shouldn't_ be false, because it should be returned by:

```
        } else if (tasks.isEmpty() == false) {
            return tasks.get(0);
        }
```

That last line of the `newResponse` method only runs if there were no responses and no errors, which seems impossible.  But I left it there just in case there is a way that can happen.  Maybe if the datafeed gets deleted in between validation of the isolate datafeed request and getting the response from it?

The alternative would be just to change the non-error response to always return `isolated: true` like acknowledged requests do.

droberts195 committed:
Address some review comments

benwtrent reviewed (approved):
:shipit:
